### PR TITLE
[Triton] DS fused custom ops

### PR DESCRIPTION
**Pull request by k50112113**
This PR contains the following:
1. added fused elementwise multiply+addtion in `aiter/ops/triton/fused_elementwise.py`
2. added fused RoPE+concat in `aiter/ops/triton/fused_concat.py`
3. added fused flatten+mxfp4 quantization and fused RMSnorm+addtion+mxfp4 quantization in `aiter/ops/triton/fused_quant.py`
4. fix "2*K" to "K" when reading `aiter/ops/triton/gemm_a16w16.py` configs
5. added ouput options and batch size checks in `aiter/ops/triton/batched_gemm_afp4wfp4_pre_quant.py`
6. added atomic add version of bf16 gemm in `aiter/ops/triton/gemm_a16w16_atomic.py`
7. updated triton mxfp4 gemm and bf16 gemm configs

### CODE REVIEW

File: aiter/ops/triton/fused_concat.py [hunk 1]
```diff
@@ -0,0 +1,425 @@
+import torch
+import triton
+import triton.language as tl
+
+
+# TODO use upstream rotation code in the future
```
Comment (rahulbatra85) [inline aiter/ops/triton/fused_concat.py]: What is this upstream code?

File: aiter/ops/triton/fused_concat.py [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+            k2_stride_h,
+            k2_stride_d,
+            k_out_stride_b,
+            k_out_stride_h,
+            k_out_stride_d,
+            BLOCK_D1,
+        )
+
+
+def fused_qk_cat(
```
Comment (rahulbatra85) [inline aiter/ops/triton/fused_concat.py]: Please add documentation for op. Add description, parameters and return values. See GEMM op for details.

File: aiter/ops/triton/fused_concat.py [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+            k_out_stride_h,
+            k_out_stride_d,
+            IS_NEOX,
+            BLOCK_D_nope,
+            BLOCK_D_pe,
+            BLOCK_D_HALF_pe,
+        )
+
+
+def fused_qk_rope_cat(
```
Comment (rahulbatra85) [inline aiter/ops/triton/fused_concat.py]: Please add documentation for op. Add description, parameters and return values. See GEMM op for details.

File: aiter/ops/triton/fused_concat.py
Comment (rahulbatra85) [inline aiter/ops/triton/fused_concat.py:1]: Please name the file that matches the kernel names. Fused_concat is too generic. We have a kernel that involves RoPE

File: aiter/ops/triton/fused_elementwise.py
Comment (rahulbatra85) [inline aiter/ops/triton/fused_elementwise.py:1]: Name the file fused_mul_add

File: aiter/ops/triton/fused_quant.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+    tl.store(
+        out_scales_ptr
+        + m * out_scales_stride_m
+        + (n1 * NUM_QUANT_BLOCKS + block_scale_offs) * out_scales_stride_n,
+        out_block_scales,
+        mask=block_scale_offs < tl.cdiv(N2, MXFP4_QUANT_BLOCK_SIZE),
+    )
+
+
+def fused_flatten_mxfp4_quant(
```
Comment (rahulbatra85) [inline aiter/ops/triton/fused_quant.py]: Please add op documentation

File: aiter/ops/triton/fused_quant.py [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+        norm2 = _rmsmorm_op(inp2, w2, inp2_n_cols, eps2)
+        tl.store(out2_ptr + pid * out2_row_stride + block_inds, norm2, mask=mask2)
+    if FIRST_INPUT_RES:
+        inp1 = inp1.to(out_res1_ptr.dtype.element_ty)
+        tl.store(
+            out_res1_ptr + pid * out_res1_row_stride + block_inds, inp1, mask=mask1
+        )
+
+
+def fused_rms_mxfp4_quant(
```
Comment (rahulbatra85) [inline aiter/ops/triton/fused_quant.py]: Please add op documentation

File: aiter/ops/triton/fused_quant.py
Comment (rahulbatra85) [inline aiter/ops/triton/fused_quant.py:1]: Please name file to represent the kernels in the file

File: aiter/ops/triton/gemm_a16w16_atomic.py
Comment (rahulbatra85) [inline aiter/ops/triton/gemm_a16w16_atomic.py:1]: Can we merge this code into gemn_a16w16.py?

**Review by rahulbatra85 — CHANGES_REQUESTED** [review]
Please see comments. Thanks!

File: aiter/ops/triton/gemm_a16w16_atomic.py
Comment (cagrikymk) [reply aiter/ops/triton/gemm_a16w16_atomic.py:1]: > Can we merge this code into gemn_a16w16.py?

Could we create a different PR to introduce k-split with reduce and atomic add options to  gemm_a16w16.py after this PR is merged? There are certain caveats with the atomic version like 0 tensor allocation (causes overhead) and potential precision loss when atomically adding to bf16/fp16 (accumulation in fp32 resolves it). So, I think it might be better to have a different PR for that so we can iron out the details there.

Comment (rahulbatra85) [reply aiter/ops/triton/gemm_a16w16_atomic.py:1]: @cagrikymk Are you suggesting to keep them split into two files for now? or take the atomic one completely out?

Comment (cagrikymk) [reply aiter/ops/triton/gemm_a16w16_atomic.py:1]: @rahulbatra85 I suggest keeping them separate for now and having another PR later to introduce atomic and reduction options to the original 16 bit GEMM kernel to simplify the code.

File: aiter/ops/triton/fused_concat.py (hunk 1 above)
Comment (k50112113) [reply aiter/ops/triton/fused_concat.py]: resolved

File: aiter/ops/triton/fused_concat.py (hunk 3 above)
Comment (k50112113) [reply aiter/ops/triton/fused_concat.py]: added

File: aiter/ops/triton/fused_elementwise.py
Comment (k50112113) [reply aiter/ops/triton/fused_elementwise.py:1]: added

File: aiter/ops/triton/fused_concat.py (hunk 2 above)
Comment (k50112113) [reply aiter/ops/triton/fused_concat.py]: added

File: aiter/ops/triton/fused_quant.py (hunk 5 above)
Comment (k50112113) [reply aiter/ops/triton/fused_quant.py]: added

File: aiter/ops/triton/fused_quant.py (hunk 4 above)
Comment (k50112113) [reply aiter/ops/triton/fused_quant.py]: added

File: aiter/ops/triton/fused_concat.py
Comment (k50112113) [reply aiter/ops/triton/fused_concat.py:1]: renamed to fused_qk_concat

File: aiter/ops/triton/fused_quant.py
Comment (k50112113) [reply aiter/ops/triton/fused_quant.py:1]: renamed to fused_mxfp4_quant
